### ISSUE TITLE
Passphrase configuration is limited

### ISSUE THREAD

**Issue by jeaye** [zulip/docker-zulip#143]
### The things which currently go wrong
* Docker Compose tries to dereference any passphrase containing `$foo` and requires `$$foo`
* The default passphrases are wrapped in double quotes, which [require escapes in YAML](https://symfony.com/doc/current/components/yaml/yaml_format.html#strings)
* If switched to single quotes, then single quotes within the passphrases need to be doubled `''`
* Single quotes, even when doubled, then fail to pass through to Postgres when it creates the initial user
* `~` isn't permitted, due to a [regex assumption](https://github.com/zulip/docker-zulip/blob/master/entrypoint.sh#L234) within `entrypoint.sh`
* Setting new passphrases requires changing a git-tracked file

### Things which can be fixed
- [ ] Passphrases should be escaped when they're sent to Postgres and other services
- [ ] The regex assumption should be replaced with something which doesn't care about the value
- [ ] The passphrases should be read from env vars and handled entirely by the user; no more changes to git-tracked files should be needed (this encourages people to commit the passphrases to git)
- [ ] We should invite and encourage people, in our docs, to use the `.env` file which docker-compose reads; docs [here](https://docs.docker.com/compose/compose-file/#variable-substitution)
- [ ] The variables used for passphrases should fail if not supplied; i.e.

```yaml
SECRETS_secret_key: "${VARIABLE:?err}"
```

**Comment by timabbott** [comment]
This fixes the regex issue 78480d4de548c6fccfc51d5434d2ad143db5f736 (and also makes the code a lot cleaner).

**Comment by timabbott** [comment]
Single quotes are obviously the better choice than double-quotes in YAML; just merged b2bb979b4ef632b2369045c4358b03af5c107671 to migrate the project.

**Comment by timabbott** [comment]
For

>￼ The variables used for passphrases should fail if not supplied; i.e.
> `SECRETS_secret_key: "${SECRETS_secret_key:?err}"`

It seems to me that we should be making a set variable that's empty fatal here:

```
    for SECRET_KEY in "${SECRETS[@]}"; do                                                                                                   
        local key="SECRETS_$SECRET_KEY"                                                                                                     
        local SECRET_VAR="${!key}"                                                                                                          
        if [ -z "$SECRET_VAR" ]; then                                                                                                       
            echo "Empty secret for key \"$SECRET_KEY\"."                                                                                    
        fi                                                                                                                                  
        crudini --set "$DATA_DIR/zulip-secrets.conf" "secrets" "${SECRET_KEY}" "${SECRET_VAR}"                                              
    done                                                                                                                                    
```

@galexrt what's the background here?

**Comment by jeaye** [comment]
`"${SECRETS_secret_key:?err}"` is fatal (as per the `?err`), as described [here](https://docs.docker.com/compose/compose-file/#variable-substitution).

**Comment by timabbott** [comment]
@jeaye for this one:
>￼ Passphrases should be escaped when they're sent to Postgres and other services

is the block that needs escaping this, or something else?
```
    export PGPASSWORD="$SECRETS_postgres_password"                                                                                          
    local TIMEOUT=60                                                                                                                        
    echo "Waiting for database server to allow connections ..."                                                                             
    while ! /usr/bin/pg_isready -h "$DB_HOST" -p "$DB_HOST_PORT" -U "$DB_USER" -t 1 >/dev/null 2>&1                                         
```

**Comment by jeaye** [comment]
Not sure anymore, but that looks likely. Adding a `"` to the middle of your passphrase should raise the issue.

**Comment by jeaye** [comment]
Glad you could repro. Seems like reporting it upstream and making it clear within docker-zulip is the best we can do for now, unless it's a quick fix. If we provided a tool for generating passphrases into `.env`, we could take care of this for the user.

**Comment by timabbott** [comment]
Agreed re: reporting upstream; can you take reporting this one?  And yeah, if we're providing a tool, we can just have the tool not use `$` and `'`.

**Comment by jeaye** [comment]
Done.

**Comment by jeaye** [comment]
https://github.com/docker-library/postgres/issues/488 has been resolved.

**Comment by timabbott** [comment]
Awesome!  I guess we need to wait for them to do a release before we can take advantage of this?

**Comment by jeaye** [comment]
Yeah, I'm not sure how that works. This was the first time I'd ever used Docker.

**Comment by timabbott** [comment]
https://hub.docker.com/_/postgres/ appears to have been last pushed to 7 hours ago, so maybe that means we can grab the latest image and we'll be OK?  Worth testing...
